test/net/sourceforge/kolmafia/textui/parsetree/ValueTest.java

```diff
@@ -0,0 +1,341 @@
+package net.sourceforge.kolmafia.textui.parsetree;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import java.util.Objects;
+import net.sourceforge.kolmafia.AscensionPath.Path;
+import net.sourceforge.kolmafia.VYKEACompanionData;
+import net.sourceforge.kolmafia.persistence.AdventureDatabase;
+import net.sourceforge.kolmafia.persistence.MonsterDatabase;
+import net.sourceforge.kolmafia.persistence.MonsterDatabase.Element;
+import net.sourceforge.kolmafia.request.DiscoGiftCoRequest;
+import net.sourceforge.kolmafia.request.NinjaStoreRequest;
+import net.sourceforge.kolmafia.textui.DataTypes;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+class ValueTest {
```

Comment:
Looking at the test in general:

When running comparisons, there are three options: x less than y, x equal to y, x greater than y. Respectively, this corresponds to a result of -1, 0 or 1. In `comparePaths`, you have `greaterThanOrEqualTo(0)` which means that you think the comparison is either equal or greater. I would hope instead you knew enough to know which one.

Somewhat similar: for `greaterThanOrEqualTo(1)`, is there any chance it is not exactly 1? Could these be "is(-1)", "is(1)", and maybe even refactored out: "isBefore()", "isAfter()"?

In general, I'd prefer the test describes what's going on: e.g. "known monsters compare by id / unknown monsters compare by string". For example, "compare effect to other effect": the assertion written presumably covers the current behaviour, but I think it should be recorded what the current behaviour is (presumably id?), to check that's being asserted against instead of the assertion just asserting that the current behaviour is right.

Great coverage though!

Reply:
> When running comparisons, there are three options: x less than y, x equal to y, x greater than y. Respectively, this corresponds to a result of -1, 0 or 1. In `comparePaths`, you have `greaterThanOrEqualTo(0)` which means that you think the comparison is either equal or greater. I would hope instead you knew enough to know which one.
> 
> Somewhat similar: for `greaterThanOrEqualTo(1)`, is there any chance it is not exactly 1? Could these be "is(-1)", "is(1)", and maybe even refactored out: "isBefore()", "isAfter()"?

It's not always -1,0 or 1, it can be any magnitude. It's the direction that matters. I agree a sign check would be preferable, but I couldn't find one and thought this was fine

Reply:
Math.signum() returns -1, 0 or 1.

Sounds like wrapping things with signum would work in this case.